hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -131,34 +187,72 @@ public PublicKey getPublicKey() {
   }
 
   /**
-   * Returns the certificate  of the specified component if it exists on the
-   * local system.
+   * Returns the default certificate of given client if it exists.
    *
    * @return certificate or Null if there is no data.
    */
   @Override
   public X509Certificate getCertificate() {
-    if(x509Certificate != null){
+    if (x509Certificate != null) {
       return x509Certificate;
     }
 
-    Path certPath = securityConfig.getCertificateLocation();
-    if (OzoneSecurityUtil.checkIfFileExist(certPath,
-        securityConfig.getCertificateFileName())) {
-      CertificateCodec certificateCodec =
-          new CertificateCodec(securityConfig);
-      try {
-        X509CertificateHolder x509CertificateHolder =
-            certificateCodec.readCertificate();
-        x509Certificate =
-            CertificateCodec.getX509Certificate(x509CertificateHolder);
-      } catch (java.security.cert.CertificateException | IOException e) {
-        getLogger().error("Error reading certificate.", e);
-      }
+    if (certSerialId == null) {
+      getLogger().error("Default certificate serial id is not set. Can't " +
+          "locate the default certificate for this client.");
+      return null;
+    }
+    // Refresh the cache from file system.
+    loadAllCertificates();
```

Comment:
Should we loadAllCertificates() again only if the map does not contain it? 

Also, in the contsructor when we call loadAllCertificates(), should we asset the passing in certSerialId is loaded into the map from file system?

Reply:
initialized local certificate during initial call to loadAllCertificates. Now if it is null at L202 than it is also not present in map, so loading it again from filesystem and checking it again is the only option.

Reply:
We can assert in constructor if certSerialId is not null than its corresponding certificate should not be null as well after loading from memory but if we fail than it will become catch 22 for some recovery scenarios. (Ex we can't even create instance of CertificateClient to call init which may handle some of the automatic recovery in future)

Reply:
Let's discuss and file follow up JIRAs. I'm OK with the current approach as-is.